Add Storage dashboard

davelopez opened the pull request:
This PR adds a new `Storage Dashboard` feature to the Galaxy client. It is my first "big" PR for the frontend so I will appreciate comments on style, best practices, patterns, and recommendations!

## General idea
The general idea is to have a **centralized place where the user can have an overview of the disk usage and quota and provide them with tools to easily recover space or manage their data**.

The current interface it's heavily inspired in [Google's storage management](https://one.google.com/storage) (maybe too heavily :sweat_smile:) by keeping it simple and guiding the user as much as possible through the options.

![StorageDashboard](https://user-images.githubusercontent.com/46503462/148061101-0c2cb696-2a9b-43fe-aee3-7e1cff258ddc.gif)

## Extensible
The Storage Manager view has been designed to be easily extended with new operations to discover and free space. Adding more cleanup operations is a matter of editing the `client/src/components/User/DiskUsage/Management/Cleanup/categories.js` and adding a new category or a new operation to an existing category like this:
```js
{
    name: _l("Old Items"), // Category name
    operations: [
        {
            id: "old_datasets", // Unique identifier for this operation
            name: _l("Old datasets"), // The display name
            description: _l("Datasets that haven't been deleted but were not modified in quite some time"),
            fetchSummary: async () => { // Callback that gets the total recoverable space and the number of items to clear
                return { totalSize: 0, totalItems: 0 };
            },
            fetchItems: async () => [], // Callback that fetches the actual items to clear
            cleanupItems: async () => {}, // Callback to purge/delete/clear a collection of items
        },
        {
            id: "old_histories",
            name: _l("Old histories"),
            description: _l("Histories that haven't been deleted but were not modified in quite some time"),
            fetchSummary: async () => {
                return { totalSize: 0, totalItems: 0 };
            },
            fetchItems: async () => [],
            cleanupItems: async () => {},
        },
    ],
},
```
This will add a new category `Old Items` with two operations to delete old datasets and histories. You just need to define proper functions for the `fetchSummary`, `fetchItems`, and `cleanupItems` callbacks.

![Screenshot from 2022-01-04 14-24-54](https://user-images.githubusercontent.com/46503462/148066511-0a98b6be-27c5-4658-bdf7-239fcfaedc2d.png)


## Open questions
- [x] **How this should work with anonymous users?** Right now this feature is not available to anonymous users because some of the API calls for retrieving datasets require a logged-in user. This probably can be changed but it's worth discussing.
   - From the discussion so far this feature makes sense only for registered users.
- [x] **What happens if the server has `allow_user_dataset_purge` disabled?** Well, right now it will throw an error when trying to permanently delete datasets. In this scenario, the "Deleted items" category may not be useful and could be conditionally hidden, we can also change the other non-deleted cleanup operations to `delete` instead of `purge` but then the quota change will not be reflected... some more material for discussion here :smile: 
   - Seems that `allow_user_dataset_purge` can be assumed to be always `true` so no issue here.

## Future work
- This PR only adds one cleanup operation, more can/will be added in different PRs as extensions.
- Add full support for multi quotas and different ObjectStores after https://github.com/galaxyproject/galaxy/pull/10977 is completed.
- In addition to the user-friendly usage cleanup utilities, we can add more options here to visually explore storage data with charts and other visualizations for power users.
- Provide more context information about a particular item when reviewing it. Maybe a popover with more details, or links to history, etc.
- Your ideas here...

## TODO (before is ready to review)
- [x] <del>Move backend changes to their own PRs to simplify review.</del> Improve and fix backend changes in this PR as Marius suggested.
- [x] Make it a bit more future-proof for https://github.com/galaxyproject/galaxy/pull/10977.
   - See https://github.com/galaxyproject/galaxy/pull/13113#issuecomment-1007598106
- [x] Add automated tests

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  - Delete some datasets
  - Click on the quota meter in the masthead or directly go to `http://127.0.0.1:8081/storage`

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

Fixes https://github.com/galaxyproject/galaxy/issues/518

davelopez committed:
Add client route to storage dashboard

davelopez committed:
Add UserDetailsProvider

davelopez committed:
Add QuotaUsageSummary component

davelopez committed:
Add QuotaSettings model

davelopez committed:
Refactor and add DiskUsageSummary component

Displays the user disk usage with and without using quotas

davelopez committed:
Add generic IconCard component

davelopez committed:
Refactor and tweak the style of components

davelopez committed:
Add StorageDashboardRouter

davelopez committed:
Add StorageManager component

The Storage manager provides a set of tools to discover posibly wasted disk space and review and delete those contents easily.

davelopez committed:
Support multiple purgeable categories

Refactor to define multiple categories and providers to display summary information about potentially purgeable items in that particular category. For example, deteled datasets, large datasets, old datasets, etc.

davelopez committed:
Add PurgableDetailsModal component

Allows to review the items that are going to be deleted and show additional confirmation.

davelopez committed:
Various refactorings and cosmetic changes

davelopez committed:
Allow custom keys serialization in datasets API

The index endpoint was not passing down to the serializer the full serialization params so it was ignoring the `keys` serialization parameter.

davelopez committed:
Display message when data cannot be immediately purged

davelopez committed:
Add API endpoint to delete/purge datasets in batch

davelopez committed:
Add API tests for datasets batch delete endpoint

davelopez committed:
Add client functions to purge multiple datasets

davelopez committed:
Display the total amount of space recovered after cleaning

- Show a dialog with the amount recovered
- Refresh the summary to display the updated values

davelopez committed:
Use prop instead of passing callback

to refresh a particular provider by name.

davelopez committed:
Cosmetic improvements

davelopez committed:
Access Storage Dashboard from Quota meter link

davelopez committed:
Refactor use CurrentUser provider

- Remove redundant UserDetailsProvider and use existing CurrentUser (provider).
- Simplify some components by removing unnecessary props.

davelopez committed:
Refactor + slightly increased scalability

- Move cleanup related files to own namespace
- Use b-table pagination with an items provider

davelopez committed:
Some UI improvements and cleaning

Add a little delay to loading spinners so they don't flicker

mvdbeek commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -160,6 +160,38 @@ class BamDataResult(DataResult):
 
+class DeleteDatasetBatchPayload(BaseModel):
+    ids: List[EncodedDatabaseIdField] = Field(
+        description="The list of datasets IDs to be deleted/purged.",
```
I think that as it is you can only delete `HistoryDatasetAssociation` items. I would add the `src` parameter (could be `hda`, `ldda`, for `DatasetInstance` subclasses -- plus `dda` soon) here to control which model you can delete.

davelopez replied:
That's a great idea. I will do that thanks!

davelopez replied:
Hopefully this is addressed in https://github.com/galaxyproject/galaxy/pull/13113/commits/3502fa15486a2dfbc6bda06864b2173fc9afadff

mvdbeek replied:
I guess the question is would we want to batch-delete mixed content ? If yes, we should accept a payload like `[{'src': 'hda'}, 'id': 1], {'src': 'ldda', 'id': 2}]` etc. It is less efficient in terms of having to repeat `src`, but on the other hand it does feel consistent with regards to other endpoints ... not sure.

davelopez replied:
Ohh I see, then it's better to make it like the other endpoints. Otherwise, it may be too tailored for this particular use case and less flexible.

mvdbeek replied:
(Just to clarify after reading again what I wrote: I should have written "other, single value payloads in other endpoints", not "other endpoints", we don't have many other batch endpoints)

mvdbeek commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -483,6 +515,29 @@ def converted(
+    ) -> DeleteDatasetBatchResult:
+        success_count = 0
+        errors = []
+        for encoded_id in payload.ids:
```
I am worried about users "destroying" their dataset collections by deleting members of their dataset collections here. That is a hard problem unfortunately. Maybe we need to split the HDAs that can be deleted into those that appear as DatasetCollectionElements and those that do not ?

mvdbeek replied:
Ah, and we should also treat HDAs that appear in published histories differently (probably in the listing step, not here. If you send the item id here we should just assume you know what you're doing).

davelopez replied:
I agree these backend changes are a bit simplistic just to get something to test with. I will open separate PRs for them and include your comments. Then we can iterate over those PRs and drop the API changes from this one once they are merged.

mvdbeek replied:
I don't particularly mind having the backend changes in here, if it's not a huge amount of refactoring it helps to see why it was changed.

jdavcs replied:
Brainstorming.. Unless this has been implemented already (looking at the model, I don't see it), we could add a field to the HDA model that stores a count of its parent collections. Whenever an HDA instance is added to a collection, we increment that field. Deleting a collection decrements the counters of its members. Deleting an HDA is only allowed when the field is 0. Like ref counts in garbage collection. 
(EDIT: i was typing this reply to the very first comment in this thread, without seeing the entire conversation :smile: )

mvdbeek replied:
That would be a field that potentially sees a lot of concurrent updates for a usecase that is probably not run that often (i suppose you're not constantly hitting the limit / worried about your storage after all).  it's not that hard to generate a query for this. If  a query turns out to be a problem we can still think about alternatives.

jdavcs replied:
Oh yes, you're right. And now that I think of it, a query shouldn't be expensive.

davelopez replied:
So, to recap and see if I understood correctly, this `delete_batch` datasets endpoint **should assume the caller knows what she/he is doing**. And is **the job of the `fetchItems` query to filter out unwanted datasets**, like "public" or those "part of a collection".
We can always add specific cleanup operations extensions targeting collections or entire histories.

mvdbeek replied:
There are pros and cons to doing it this way.
The downside is that in between fetchItems and calling `delete_batch` the user may have changed something (published a history, added a HDA to a collection) and now their deleting things that should have probably not been deleted ... but that doesn't seem like a very likely scenario.

The pro is that it is simpler and a good enough start IMO.
One way to give stronger guarantees about not deleting unwanted items is to build an atomic update query where you basically use the same query that you'd use to get a list of items that is not in a published history / part of a collection and you turn that into the update statement. That would probably be a significant amount of additional effort.

My gut feeling is that as long as we document in the delete_batch endpoint that no additional checks (beyond permission) are performed that this is good enough, and we could add a "strict" mode later if it becomes necessary.

mvdbeek replied:
(but if you want to implement the "strict" mode or another way to do something like this now that's cool too, it is your call)

mvdbeek reviewed (commented):
A quick review here. This looks really good for as far as I can judge the client portion (I'd have used a ConfigProvider instead of the explicit call to `getGalaxyInstance`, but it still looks very clean).

jdavcs commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -200,3 +201,61 @@ def test_update_datatype(self):
+
+        assert deleted_result["success_count"] == len(expected_deleted_ids)
+        for purged_id in expected_deleted_ids:
+            dataset = self._get(f"histories/{history_id}/contents/{purged_id}").json()
```
```suggestion
        for deleted_id in expected_deleted_ids:
            dataset = self._get(f"histories/{history_id}/contents/{deleted_id}").json()
```
Assuming that was the intention?

davelopez replied:
It was! thanks! :)

davelopez committed:
Update lib/galaxy_test/api/test_datasets.py

Co-authored-by: Sergey Golitsynskiy <sgolitsynskiy@gmail.com>

davelopez committed:
Remove QuotaSettings wrapper

Better to directly use the ConfigProvider and simplify a bit the code.

davelopez committed:
Add API test for custom key serialization

Tests that the datasets `index` endpoint can return specific dataset fields.

davelopez committed:
Add `src` field to DeleteDatasetBatchPayload

This allows to delete/purge different kinds of dataset associations.
All datasets will be expected to have the same type.

mvdbeek commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -160,6 +160,43 @@ class BamDataResult(DataResult):
+
+
+class DatasetErrorMessage(BaseModel):
+    dataset_id: EncodedDatabaseIdField = Field(
```
A general remark here: Most of the time (but not all, unfortunately) when we write `dataset_id` we are referring to the `id` column of the `dataset` table. "Datasets" as they appear in histories or libraries are always `{History,LibraryDataset}DatasetAssociation`s, which have a `dataset_id` column that points to the id of the Dataset instance, and in some API endpoints we expose this dataset_id. So if we say `dataset_id` here it isn't clear what this refers to.

My suggestion would be to call this field `id` and to add `src` here as well. The client can then translate this into a a more user-friendly error using these two fields.

davelopez replied:
Thanks! I care a lot about correct naming, but it's usually the hardest part :laughing: 
I changed some of the namings in https://github.com/galaxyproject/galaxy/pull/13113/commits/508926f034a6873cf2dc5836938c18ea654e5bbf, but not sure if they are more clarifying. This one in particular I changed it to `dataset` because now is not just the `id` but also the `src` and is somewhat like a "dataset reference".

Happy to change them further now that we are at it.

mvdbeek replied:
That looks good, thank you!

davelopez committed:
Improve result error handling

Show error details in result dialog when the cleanup request fails.

davelopez committed:
Add `src` for each dataset `id` when batch deleting

Allows to delete mixed types of datasets (currently `hda` or `ldda`) in the same request

davelopez committed:
Improve cleanup result dialog error handling

Display the number of failed datasets if some of them couldn't be deleted

mvdbeek commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -160,6 +161,38 @@ class BamDataResult(DataResult):
+    errors: Optional[List[DatasetErrorMessage]] = Field(
+        default=None,
+        description=(
+            "A collection of dataset IDs and the corresponding error message if something "
```
```suggestion
            "A list of dataset IDs and the corresponding error message if something "
```
Just to reserve `collection` for dataset collections, even if that is clear here.

davelopez committed:
Improve code documentation

Based on Marius' comments

davelopez committed:
Add wait for new dataset to be created in test

`test_index_using_keys` doesn't seem to fail on local but fails on CI because it returns an empty object instead of the object with the size.
I wonder if this is because the dataset is not fully populated when checking and hopefully this will fix it.

davelopez committed:
Update lib/galaxy/webapps/galaxy/services/datasets.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Refactor move cleanup models into cleanup namespace

davelopez committed:
Add `quota_bytes` field to user serialization in API

The value in bytes is more useful to control the final representation

davelopez committed:
Add support for multiple quota sources

This will hopefully simplify future integration with https://github.com/galaxyproject/galaxy/pull/10977

davelopez committed:
Add basic tests for DiskUsageSummary

davelopez committed:
Add tests for QuotaUsage model

davelopez committed:
Add tests for QuotaUsageBar component

davelopez committed:
Add tests for QuotaUsageSummary component

davelopez committed:
Center spinner when loading the quota usage summary

davelopez committed:
Fix test_index_using_keys

For some reason this test keeps returning {} (empty object) instead of the `size` when running on CI (but passes on local).
Maybe this `size` key gets populated after some time?
I changed the expected key to `id` and check there is only one key returned which is what the test is really about.

davelopez committed:
Simplify CleanupResultDialog

and fix partial succes use case where some items were cleared and some others were not

davelopez committed:
Replace ResourceCleanupManager with CleanupCategoriesProvider

davelopez committed:
Fix partial success condition

davelopez committed:
Add better error reporting for partial success operations

davelopez committed:
Allow to select all item through all pages

This will fetch and select all cleanable items without pagination.

davelopez committed:
Mark the Storage Manager as Beta

davelopez committed:
Add tests for CleanupResultDialog

davelopez committed:
Refactor CleanOperationSummary

- Make refresh delay configurable

davelopez committed:
Apply only minimum refresh delay for visual feedback

Only delay if the minimum delay was not reached by the fetch summary callback

davelopez committed:
Fix cleanup operation card content alignment

The card content is now properly aligned between cards in the same category.

davelopez committed:
Add tests for CleanupOperationSummary component

davelopez committed:
Add more documentation for CleanupOperation

davelopez committed:
Fix async handling in test (hopefully)

davelopez committed:
Allow displaying ReviewCleanupDialog from show prop

davelopez committed:
Make total number of items to review more prominent

davelopez committed:
Add ReviewCleanupDialog tests

+ small refactoring

davelopez committed:
Add Beta warning to the Storage Manager

davelopez committed:
Add disk usage recalculation button

This will trigger a disk usage recalculation in the backend and display some vague feedback since we don't really know when it will be ready.

davelopez committed:
Remove disk usage recalculation from logout

SInce this now can be triggered from the Storage Dashboard

mvdbeek commented on client/src/components/User/DiskUsage/Quota/QuotaUsageProvider.js:
```diff
@@ -0,0 +1,24 @@
+ * Fetches the disk usage by the user across all ObjectStores.
+ * @returns {Array<QuotaUsage>}
+ */
+async function fetchQuotaUsage() {
```
The user quota has a modification time. You could use that to check if the update is complete. But maybe that's premature optimization. Not needed for a first pass.

davelopez replied:
Ideally, we could get the updated quota value notified from a Vuex store right? Also after the celery task for quota recalculation finishes, I need to get more familiar with those stores yet :sweat_smile: 
So yes, it looks like a good follow up optimization :)

mvdbeek commented on lib/galaxy/webapps/galaxy/controllers/user.py:
```diff
@@ -252,19 +251,6 @@ def logout(self, trans, logout_all=False, **kwd):
         message = trans.check_csrf_token(kwd)
         if message:
             return self.message_exception(trans, message)
-        if trans.user:
```
You'll want to update https://github.com/galaxyproject/galaxy/blob/dev/client/src/mvc/user/user-quotameter.js#L149 as well

mvdbeek commented on client/src/mvc/user/user-quotameter.js:
```diff
@@ -134,11 +135,11 @@ var UserQuotaMeter = Backbone.View.extend(baseMVC.LoggableMixin).extend(
                 : "";
             const using = `${_l("Using")} ${data.quota_percent}%`;
-            const quotaUrl = this.options.quotaUrl;
+            const storageDashboardUrl = `${getAppRoot()}storage`;
```
I think the dashboard should always be reachable when you click on the quota

mvdbeek commented on client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue:
```diff
@@ -0,0 +1,252 @@
+                },
+                {
+                    key: "size",
+                    sortable: false, // The API does not allow sort by size?
```
We should add this, without sorting this will not work very well.

mvdbeek replied:
4f6ff859f4ca7a49740a08f3a70219a78c376b9c works for me, but we should add some test there

davelopez replied:
Thank you so much for the fix! I will try to add some tests too :+1:

mvdbeek commented on client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue:
```diff
@@ -0,0 +1,252 @@
+            ],
+            sortBy: "update_time",
+            sortDesc: true,
+            perPage: 5,
```
That is too small to be usable IMO
```suggestion
            perPage: 50,
```

davelopez replied:
Ups... this is a leftover from testing the pagination, thanks for the catch!

mvdbeek committed:
Enable sorting by dataset size

mvdbeek commented on client/src/components/User/DiskUsage/Management/Cleanup/ReviewCleanupDialog.vue:
```diff
@@ -0,0 +1,252 @@
+        <div>
+            {{ captionText }}
+            <b>
+                <b-link @click="onSelectAllItems">select all {{ totalItems }} items</b-link>
```
total items maxes out at 500 because of the API limit, there may be more items actually. Not sure how to convery this, but if we sort by size it's maybe not that important

davelopez replied:
Oh, you are right, I didn't try with so many datasets. I guess we can set the `limit` query parameter in the requests to a pretty higher number like 10000 to bypass the 500 default limit but still there will be a limit.

The sorting by size seems to be working pretty well though :)

mvdbeek commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -493,6 +529,36 @@ def converted(
+                if dataset.src == DatasetSourceType.hda:
+                    self.hda_manager.error_if_uploading(dataset_instance)
+                if payload.purge:
+                    manager.purge(dataset_instance, flush=False)
```
The purge especially could take a while. I think this should ideally be an asynchronous celery endpoint (in a follow up).

davelopez replied:
:+1: added to my TODO list for follow-ups

mvdbeek reviewed (commented):
This is working well for me, I'm looking forward to the remaining categories :).

davelopez committed:
Fix wrong number of items per page

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Make the dashboard always reachable

- Now it can be reached also when no quota limit is assigned to the user
- Remove the quota recalculation after login tooltip

davelopez committed:
Add API test for dataset size ordering

davelopez committed:
Increase discarded items fetch limit

Set to 5000 to bypass the default 500 limit, but still limited.

mvdbeek commented on client/src/components/User/DiskUsage/Management/services.js:
```diff
@@ -0,0 +1,119 @@
+const isDataset = "q=history_content_type-eq&qv=dataset";
+const isDeleted = "q=deleted-eq&qv=True";
+const isNotPurged = "q=purged-eq&qv=False";
+const maxItemsToFetch = 5000;
```
I think that can be very heavy on the server. I'd either leave this unspecified and adjust the message to say "more than 500 items" or I'd limit this at a lower number and return the total matches in the header.

davelopez replied:
Understand!, I will limit it back to 500 and indicate this limit in the operation description so the user doesn't get confused if more items appear after the first pass. Since it's ordered by size by default it should be good enough to review/bulk delete in chunks of 500.

davelopez replied:
Ok, the limit is now back to 500 and the review dialog will display `500+` and a clarification message in the footer when there are at least 500 items

![Screenshot from 2022-03-25 13-15-15](https://user-images.githubusercontent.com/46503462/160119657-f600740b-9bfc-44b3-b366-30cc24e129ff.png)

davelopez committed:
Set fetch limit back to 500 and display limit indication
